In [1]:
from openai import OpenAI

In [2]:
import os
from dotenv import load_dotenv
load_dotenv()
openai_key = os.getenv("OPENAI_API_KEY")
print(os.environ['OPENAI_API_KEY'])

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


In [3]:
from openai import OpenAI
client = OpenAI() 

In [4]:
assistant = client.beta.assistants.create(
    name="Mortgage Bot",
    instructions="You use Python code to help answer questions about mortgage and interest payments.",
    tools=[{'type':'code_interpreter'}],
    model = 'gpt-3.5-turbo',
)

In [5]:
thread = client.beta.threads.create()

In [6]:
thread.id

'thread_g4Bh5O1m6Q9f7Bnme56mkWe2'

In [7]:
message = client.beta.threads.messages.create(
    thread_id=thread.id,
    role='user',
    content='I want to buy a house that costs $2.1 Million on a 30 year mortgage with a 3.5% interest rate. How much will my monthly payment be?',
)

In [8]:
run = client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id,
    instructions="Give a detailed analysis and considerations regarding the mortgage payments.",
)

In [9]:
import time

while run.status != 'completed':
    run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)
    print(run.status)
    time.sleep(5)

queued
in_progress
completed


In [10]:
messages = client.beta.threads.messages.list(thread_id=thread.id, order='asc', after=message.id)

In [11]:
#messages

In [12]:
def dispaly_thread_messages(messages):
    for data in messages.data:
        print(data.content[0].text.value)
        print('\n')

In [13]:
dispaly_thread_messages(messages)

To calculate the monthly mortgage payment, we can use the formula for a fixed-rate mortgage:

\[ M = P \cdot \frac{r(1+r)^n}{(1+r)^n - 1} \]

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal amount (loan amount)
- \( r \) is the monthly interest rate (annual interest rate divided by 12 months)
- \( n \) is the total number of payments (loan term in years multiplied by 12 months)

Given:
- Principal amount (\( P \)) = $2.1 Million
- Annual interest rate = 3.5%
- Loan term = 30 years

Let's calculate the monthly mortgage payment according to the given information:


The monthly mortgage payment for a $2.1 Million house on a 30-year mortgage with a 3.5% interest rate would be approximately $9,429.94.




In [14]:
def wait_on_run(run,thread):
    while run.status == 'queued' or run.status == 'in_progress':
        run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)
        time.sleep(5)
        print(run.status)
        
    return run     

In [15]:
thread.id

'thread_g4Bh5O1m6Q9f7Bnme56mkWe2'

In [17]:
#client.beta.threads.messages.list(thread_id=thread.id)

SyncCursorPage[Message](data=[Message(id='msg_D4ziI6K8FLm6saZJNbwipijk', assistant_id='asst_fHjANrKXYsYweLuwwjofbwAp', attachments=[], completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='The monthly mortgage payment for a $2.1 Million house on a 30-year mortgage with a 3.5% interest rate would be approximately $9,429.94.'), type='text')], created_at=1722187116, incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_08TtiL3m2LZodeAnb8VuGAlS', status=None, thread_id='thread_g4Bh5O1m6Q9f7Bnme56mkWe2'), Message(id='msg_8gNZz2e7wYTzETE0yIaHrt0J', assistant_id='asst_fHjANrKXYsYweLuwwjofbwAp', attachments=[], completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="To calculate the monthly mortgage payment, we can use the formula for a fixed-rate mortgage:\n\n\\[ M = P \\cdot \\frac{r(1+r)^n}{(1+r)^n - 1} \\]\n\nWhere:\n- \\( M \\) is the monthly mortgage payment\n- \\( P \\) is the princi

In [18]:
message = client.beta.threads.messages.create(thread_id=thread.id, 
                                              role='user', 
                                              content='What about if i put down payment of $200k on the house?, how would that change my monthly payment')

In [19]:
client.beta.threads.messages.list(thread_id=thread.id)

SyncCursorPage[Message](data=[Message(id='msg_8lBBefhuyECf2h9Eybjd5low', assistant_id=None, attachments=[], completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='What about if i put down payment of $200k on the house?, how would that change my monthly payment'), type='text')], created_at=1722187419, incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='user', run_id=None, status=None, thread_id='thread_g4Bh5O1m6Q9f7Bnme56mkWe2'), Message(id='msg_D4ziI6K8FLm6saZJNbwipijk', assistant_id='asst_fHjANrKXYsYweLuwwjofbwAp', attachments=[], completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='The monthly mortgage payment for a $2.1 Million house on a 30-year mortgage with a 3.5% interest rate would be approximately $9,429.94.'), type='text')], created_at=1722187116, incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_08TtiL3m2LZodeAnb8VuGAlS', status=

In [20]:
run = client.beta.threads.runs.create(thread_id=thread.id, 
                                      assistant_id=assistant.id, 
                                      )
wait_on_run(run,thread)

in_progress
completed


Run(id='run_f0cSDJDrmc0Ygc7Uoquo4IyW', assistant_id='asst_fHjANrKXYsYweLuwwjofbwAp', cancelled_at=None, completed_at=1722187719, created_at=1722187716, expires_at=None, failed_at=None, incomplete_details=None, instructions='You use Python code to help answer questions about mortgage and interest payments.', last_error=None, max_completion_tokens=None, max_prompt_tokens=None, metadata={}, model='gpt-3.5-turbo', object='thread.run', required_action=None, response_format='auto', started_at=1722187716, status='completed', thread_id='thread_g4Bh5O1m6Q9f7Bnme56mkWe2', tool_choice='auto', tools=[CodeInterpreterTool(type='code_interpreter')], truncation_strategy=TruncationStrategy(type='auto', last_messages=None), usage=Usage(completion_tokens=192, prompt_tokens=1847, total_tokens=2039), temperature=1.0, top_p=1.0, tool_resources={}, parallel_tool_calls=True)

In [24]:
mesages = client.beta.threads.messages.list(thread_id=thread.id, order='asc', after=message.id)

In [25]:
dispaly_thread_messages(messages) 

To calculate the monthly mortgage payment, we can use the formula for a fixed-rate mortgage:

\[ M = P \cdot \frac{r(1+r)^n}{(1+r)^n - 1} \]

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal amount (loan amount)
- \( r \) is the monthly interest rate (annual interest rate divided by 12 months)
- \( n \) is the total number of payments (loan term in years multiplied by 12 months)

Given:
- Principal amount (\( P \)) = $2.1 Million
- Annual interest rate = 3.5%
- Loan term = 30 years

Let's calculate the monthly mortgage payment according to the given information:


The monthly mortgage payment for a $2.1 Million house on a 30-year mortgage with a 3.5% interest rate would be approximately $9,429.94.




To calculate the monthly mortgage payment, we can use the formula for a fixed-rate mortgage:

\[ M = P \cdot \frac{r(1+r)^n}{(1+r)^n - 1} \]

Where:
- \( M \) is the monthly mortgage payment
- \( P \) is the principal amount (loan amount)
- \( r \) is the monthly interest rate (annual interest rate divided by 12 months)
- \( n \) is the total number of payments (loan term in years multiplied by 12 months)

Given:
- Principal amount (\( P \)) = $2.1 Million
- Annual interest rate = 3.5%
- Loan term = 30 years

Let's calculate the monthly mortgage payment according to the given information:


The monthly mortgage payment for a $2.1 Million house on a 30-year mortgage with a 3.5% interest rate would be approximately $9,429.94.



In [26]:
run_steps = client.beta.threads.runs.steps.list(thread_id=thread.id, run_id=run.id)

In [29]:
for step in run_steps.data:
    print(step.step_details)

MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_qb9NXEE2XNrGcnxuP6OSiEAd'), type='message_creation')
ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_qb78vp9WbhUkGWmDKyzM0JBt', code_interpreter=CodeInterpreter(input='# New principal amount after down payment\nnew_principal_amount = principal_amount - 200000  # $200k down payment\n\n# Calculate new monthly mortgage payment using the formula with the new principal amount\nnew_monthly_payment = new_principal_amount * (monthly_interest_rate * (1 + monthly_interest_rate)**total_payments) / ((1 + monthly_interest_rate)**total_payments - 1)\nnew_monthly_payment', outputs=[]), type='code_interpreter')], type='tool_calls')
MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_LLQ2nJrd00Mo7eMxLMkK1K8T'), type='message_creation')


In [30]:
response = client.beta.assistants.delete(assistant.id)

In [31]:
response

AssistantDeleted(id='asst_fHjANrKXYsYweLuwwjofbwAp', deleted=True, object='assistant.deleted')